torchtune/modules/vision_transformer.py

```diff
@@ -0,0 +1,334 @@
+            >>> # torchtune.models.clip._position_embeddings.clip_vision_encoder
+            >>> # model = VisionTransformer(
+            ... #           indices_return_hidden = [1,2,3,4,5],
+            ... #           patch_size = 40,
+            ... #           patch_grid_size = patch_grid_size,
+            ... #           embed_dim = 32,
+            ... #           num_layers = 6,
+            ... #           in_channels = num_channels,
+            ... #           ...)
+            >>>
+            >>> x, hidden_states = model(images = batch_image, aspect_ratio = batch_aspect_ratio)
+            >>>
+            >>> # (bsz, num_tiles, num_patches_per_tile + CLS token, embed_dim)
+            >>> print(x.shape)
+            torch.Size([1, 2, 101, 32]
+            >>>
+            >>> # (bsz, len(indices_return_hidden), num_tiles, num_patches_per_tile + CLS token, embed_dim)
+            >>> print(hidden_states.shape)
+            torch.Size([1, 5, 2, 101, 32]
+        """
+        hidden_states = []
+
+        # parse inputs
+        if images.ndim == 4:
+            n_tiles = 1
+            bsz, nch, w, h = images.shape
+            assert (
+                aspect_ratio is None or aspect_ratio.max() == 1
+            ), f"since images.ndim == 4, aspect ratio must be none or 1x1, got {aspect_ratio}"
+
+        elif images.ndim == 5:
+            bsz, n_tiles, nch, w, h = images.shape
+
+        else:
+            raise ValueError(
+                f"Unsupported number of dimensions: {images.ndim}. Expected 4 or 5."
+            )
+
+        # if aspect_ratio is not provided, it defaults to one tile [1,1]
+        if aspect_ratio is None:
+            aspect_ratio = torch.tensor([[1, 1]] * bsz, device=images.device)
+
+        images = images.reshape(bsz * n_tiles, nch, w, h)
+        aspect_ratio = aspect_ratio.reshape(bsz, 2)
+
+        # patch embeddings (tokens)
+        # out: (bsz, embed_dim, patch_grid_size, patch_grid_size)
+        x = self.conv(images)
+
+        # out: (bsz, patch_grid_size**2, embed_dim)
+        x = x.flatten(start_dim=2).permute(0, 2, 1)
+
+        _, n_tokens, embed_dim = x.shape
+
+        # pre_tile_pos_embed
+        if self.pre_tile_pos_embed:
+            x = x.reshape(bsz, n_tiles, n_tokens, embed_dim)
+            x = self.pre_tile_pos_embed(x, aspect_ratio)
+
+        # apply cls token
+        x = x.reshape(bsz * n_tiles, n_tokens, embed_dim)
+        x = self.cls_token_embedding(x)
+        n_tokens += 1
+
+        # token_pos_embedding
+        x = x.reshape(bsz, n_tiles, n_tokens, embed_dim)
+        x = self.token_pos_embedding(x, aspect_ratio)
+
+        # norm
+        x = self.ln_pre(x)
+
+        # transformer with optional hidden layer outputs
+        x = x.view(bsz, n_tiles * n_tokens, embed_dim)
+        for layer_idx, transformer_layer in enumerate(self.transformer_layers):
+            if self.indices_return_hidden and layer_idx in self.indices_return_hidden:
```

Comment:
good catch! Philip and I initially got confused too, but this is expected and it is in the docstring. 

The reasoning is that you dont need to return the last hidden state activation, because thats the output itself. And you may want the input layer_idx=0, which you can retrieve it like this.